File: nelman-kmcLATTICEMODEL/source/EnergyMap.cpp
#include <math.h>

#include <iostream>
#include <stdio.h>
#include <fstream>
#include <sstream>
 
#include "EnergyMap.hpp"
#include "Stats.hpp"
#include "Lattice.hpp"
#include "Native.hpp"
#include "ptemp.h"
#include "App.hpp"
//#define AGGR   


EnergyMap::EnergyMap(Lattice * l1){
  l=l1;
  for(int i=0;i<MAX_PROCS;i++){
    for(int j=0;j<QMAX;j++){ 
      energyMap[i][j]=0;
    }
  }
}

void EnergyMap::setBetaID(int bID){
  betaID=bID;
}

int lowestEfound=8000;
void EnergyMap::mapStats(Stats & stats, double boltz){
  if(!App::nativeStats){
    if(stats.Cext < 0 || stats.Cext >=QMAX){
      cout<<"ERROR too many (" << stats.Cext << ") external contacts"<<endl;
      exit(1);
    }
    if(isnan(boltz)){
      cout<<"ERROR boltz nan"<<endl;
      exit(1);
    }
    energyMap[betaID][stats.Cext]+= boltz;
  }else{
    //check if within limits
    if(stats.Nint < 0 || stats.Nint >=QMAX){
      cout<<"ERROR too many (" << stats.Nint << ") native contacts"<<endl;
      exit(1);
    }
    // check if we need to find a lowest energy structure
    if(App::checkBelowNative && l->native->isBelowNativeEnergy(stats.Etot)){
      checkBelowNative(stats, boltz);
    }
    // map stats, depending on which type is requested
    if(App::cintStats){
      energyMap[betaID][stats.Cint] += boltz;
    }else{
      energyMap[betaID][stats.Nint] += boltz;
    }
  }
  // if collecting hbond stats, add these
  if(App::hbondStats){
    hbondMap[betaID][stats.Hext] +=boltz;
  }
  //add energy (for calculation CV)
  eMap[betaID][stats.Etot] +=boltz;
}

void EnergyMap::initTableStats(){
  std::stringstream fnStats;
  fnStats  << "statsTable"<<myRank <<".txt"; 
  tableStatsStream.open(fnStats.str().c_str());
  
  tableStatsStream<<"#step"<<"\t";
  tableStatsStream<<"T"<<"\t";

  tableStatsStream<<"beta"<<"\t";
  // tableStatsStream <<"boltz"<<"\t";
  
  tableStatsStream<<"weight"<<"\t";

  tableStatsStream <<"Eint"<<"\t";
  tableStatsStream <<"Eext"<<"\t";
  tableStatsStream <<"Esol"<<"\t";
  tableStatsStream <<"Etot"<<"\t";

  tableStatsStream <<"Cint"<<"\t";
  tableStatsStream <<"Cext"<<"\t";
  tableStatsStream <<"Ctot"<<"\t";

 if(App::nativeStats){
    tableStatsStream <<"Nint"<<"\t";
    tableStatsStream <<"Next"<<"\t";
    tableStatsStream <<"Ntot"<<"\t";
  }
  
  if(App::hbondStats){
    tableStatsStream <<"Hint"<<"\t";
    tableStatsStream <<"Hext"<<"\t";
    tableStatsStream <<"Htot"<<"\t";
  }

  if(App::eBetaMu>0 || App::clustStats){
    tableStatsStream <<"eBM"<<"\t";
    tableStatsStream <<"clust"<<"\t";
    tableStatsStream <<"maxCl"<<"\t";
    tableStatsStream <<"nChains"<<"\t";
  }

  tableStatsStream <<endl;
}

void EnergyMap::writeTableStats(Stats & stats,int step,double weight, ClusterInfo ci){
  double beta = betas[betaID]*100;
  std::stringstream buf;
  buf << std::fixed << std::setprecision(4) << 1/beta;  

  tableStatsStream<<step<<"\t";
  tableStatsStream<<buf.str()<<"\t";
  
  tableStatsStream <<beta <<"\t";
  //tableStatsStream <<boltz<<"\t";

  tableStatsStream <<weight <<"\t";

  tableStatsStream <<stats.Eint<<"\t";
  tableStatsStream <<stats.Eext<<"\t";
  tableStatsStream <<stats.Esol<<"\t";
  tableStatsStream <<stats.Etot<<"\t";

  tableStatsStream <<stats.Cint<<"\t";
  tableStatsStream <<stats.Cext<<"\t";
  tableStatsStream <<stats.Ctot<<"\t";

  if(App::nativeStats){
    tableStatsStream <<stats.Nint<<"\t";
    tableStatsStream <<stats.Next<<"\t";
    tableStatsStream <<stats.Ntot<<"\t";
  }
  
  if(App::hbondStats){
    tableStatsStream <<stats.Hint<<"\t";
    tableStatsStream <<stats.Hext<<"\t";
    tableStatsStream <<stats.Htot<<"\t";
  }
  if(App::eBetaMu>0 || App::clustStats){
    tableStatsStream <<App::eBetaMu<<"\t";
    tableStatsStream <<ci.clusters<<"\t";
    tableStatsStream <<ci.largestCluster<<"\t";
    tableStatsStream <<ci.totalChains<<"\t"; 
  }

  tableStatsStream <<endl;
}
void EnergyMap::finalizeTableStats(){
  tableStatsStream.close();
}


void EnergyMap::checkBelowNative(Stats & stats, double boltz){
  if(App::findLowestE){
    if(lowestEfound > stats.Etot){
      cout<< "energy below native energy"<<endl;
      cout<< "Mapped stats:"<<endl;
      stats.printCout();
      cout<< "Lattice stats:"<<endl;
      l->stats.printCout();
      if(l->stats.Etot == stats.Etot){
	std::stringstream outPDB;
	outPDB << "lowestEnergyN"<<myRank <<".pdb"; 
	l->writePDBMultiChain(outPDB.str());
	lowestEfound=stats.Etot;
	std::stringstream outSTATS;
	outSTATS  << "lowestEnergyN"<<myRank <<".txt"; 
	ofstream infoFile;
	infoFile.open(outSTATS.str().c_str());  
	infoFile<<stats.print2string()<<endl; 
	infoFile << "beta " << l->getBetaMoves() <<endl;
	infoFile.close();
      }
    }
  }else{
    if(! l->native->hasNativeStructure(stats.Ntot) ){ 
      cout<<"ERROR energy below native energy"<<endl;
      cout<< "Mapped stats:"<<endl;
      stats.printCout();
      cout<< "Lattice stats:"<<endl;
      l->stats.printCout();
      if(l->stats.Etot == stats.Etot){
	l->writePDBMultiChain("belowNativeEnergy.pdb");
	exit(1);
      }   
    }
  }
}

void EnergyMap::printGnuPlotDataWR(string fn_stats,int num_procs, int myRank){
  ofstream energyFile;
  energyFile.open(fn_stats.c_str());
  cout << "printing to "<<fn_stats <<endl;

  for(int betaID=0;betaID<num_procs;betaID++){ 
    energyFile<<"BETA\t"<<100.0 *  betas[betaID] << endl;
    
    for(int q=0;q<QMAX;q++){
      if(energyMap[betaID][q]>0){
	energyFile<< q <<"\t"<< energyMap[betaID][q] <<endl;
      }
    }
    //energyFile.close();
  }
  energyFile.close();
  cout << "finished printing to "<<fn_stats <<endl;
}




void EnergyMap::printEMap(string fn_stats,int num_procs, int myRank){
  
  ofstream energyFile;
  energyFile.open(fn_stats.c_str());
  cout << "printing to "<<fn_stats <<endl;
  for(int betaID=0;betaID<num_procs;betaID++){
    energyFile<<"BETA\t"<<100.0 *  betas[betaID] << endl;
    //__gnu_cxx::hash_map<int,double>::iterator iter;
    tr1::unordered_map<int,double>::iterator iter;
    for( iter = eMap[betaID].begin(); iter != eMap[betaID].end(); iter++ ) {
      energyFile<< (double) iter->first / 100.0 <<"\t"<< iter->second<<endl;
    }
  }
  energyFile.close();
}



void EnergyMap::printHBondMap(string fn_stats,int num_procs, int myRank){
  ofstream energyFile;
  energyFile.open(fn_stats.c_str());
  cout << "printing to "<<fn_stats <<endl;

  for(int betaID=0;betaID<num_procs;betaID++){ 
    energyFile<<"BETA\t"<<100.0 *  betas[betaID] << endl;
    
    for(int q=0;q<QMAX;q++){
      if(hbondMap[betaID][q]>0){
	energyFile<< q <<"\t"<< hbondMap[betaID][q] <<endl;
      }
    }
    //energyFile.close();
  }
  energyFile.close();
  cout << "finished printing to "<<fn_stats <<endl;
}

File: nelman-kmcLATTICEMODEL/source/EnergyMap.py
from typing import Dict, List
import math
from Stats import Stats
from Lattice import Lattice

QMAX = 10000
MAX_PROCS = 100  # Adjust based on your needs

class ClusterInfo:
    def __init__(self):
        self.clusters = 0
        self.largestCluster = 0
        self.totalChains = 0

class EnergyMap:
    def __init__(self, lattice: Lattice):
        self.l = lattice
        self.energy_map = [[0.0 for _ in range(QMAX)] for _ in range(MAX_PROCS)]
        self.hbond_map = [[0.0 for _ in range(QMAX)] for _ in range(MAX_PROCS)]
        self.e_map = [dict() for _ in range(MAX_PROCS)]  # Using dict instead of unordered_map
        self.beta_id = 0
        self.table_stats_stream = None
        self.lowest_efound = 8000
    
    def setBetaID(self, b_id: int):
        self.beta_id = b_id
    
    def mapStats(self, stats: Stats, boltz: float):
        if not App.nativeStats:
            if stats.Cext < 0 or stats.Cext >= QMAX:
                print(f"ERROR too many ({stats.Cext}) external contacts")
                sys.exit(1)
            if math.isnan(boltz):
                print("ERROR boltz nan")
                sys.exit(1)
            self.energy_map[self.beta_id][stats.Cext] += boltz
        else:
            if stats.Nint < 0 or stats.Nint >= QMAX:
                print(f"ERROR too many ({stats.Nint}) native contacts")
                sys.exit(1)
            
            if App.checkBelowNative and self.l.native.isBelowNativeEnergy(stats.Etot):
                self.checkBelowNative(stats, boltz)
            
            if App.cintStats:
                self.energy_map[self.beta_id][stats.Cint] += boltz
            else:
                self.energy_map[self.beta_id][stats.Nint] += boltz
        
        if App.hbondStats:
            self.hbond_map[self.beta_id][stats.Hext] += boltz
        
        self.e_map[self.beta_id][stats.Etot] = self.e_map[self.beta_id].get(stats.Etot, 0) + boltz
    
    def initTableStats(self):
        filename = f"statsTable{App.myRank}.txt"
        self.table_stats_stream = open(filename, 'w')
        headers = [
            "#step", "T", "beta", "weight",
            "Eint", "Eext", "Esol", "Etot",
            "Cint", "Cext", "Ctot"
        ]
        
        if App.nativeStats:
            headers.extend(["Nint", "Next", "Ntot"])
        if App.hbondStats:
            headers.extend(["Hint", "Hext", "Htot"])
        if App.eBetaMu > 0 or App.clustStats:
            headers.extend(["eBM", "clust", "maxCl", "nChains"])
        
        self.table_stats_stream.write("\t".join(headers) + "\n")
    
    def writeTableStats(self, stats: Stats, step: int, weight: float, ci: ClusterInfo):
        beta = self.l.betas[self.beta_id] * 100
        temp = 1 / beta if beta != 0 else float('inf')
        
        data = [
            str(step), f"{temp:.4f}", str(beta), str(weight),
            str(stats.Eint), str(stats.Eext), str(stats.Esol), str(stats.Etot),
            str(stats.Cint), str(stats.Cext), str(stats.Ctot)
        ]
        
        if App.nativeStats:
            data.extend([str(stats.Nint), str(stats.Next), str(stats.Ntot)])
        if App.hbondStats:
            data.extend([str(stats.Hint), str(stats.Hext), str(stats.Htot)])
        if App.eBetaMu > 0 or App.clustStats:
            data.extend([
                str(App.eBetaMu), str(ci.clusters),
                str(ci.largestCluster), str(ci.totalChains)
            ])
        
        self.table_stats_stream.write("\t".join(data) + "\n")
    
    def finalizeTableStats(self):
        if self.table_stats_stream:
            self.table_stats_stream.close()
    
    def checkBelowNative(self, stats: Stats, boltz: float):
        if App.findLowestE:
            if self.lowest_efound > stats.Etot:
                print("energy below native energy")
                print("Mapped stats:")
                stats.printCout()
                print("Lattice stats:")
                self.l.stats.printCout()
                
                if self.l.stats.Etot == stats.Etot:
                    filename = f"lowestEnergyN{App.myRank}.pdb"
                    self.l.writePDBMultiChain(filename)
                    self.lowest_efound = stats.Etot
                    
                    info_filename = f"lowestEnergyN{App.myRank}.txt"
                    with open(info_filename, 'w') as f:
                        f.write(stats.print2string() + "\n")
                        f.write(f"beta {self.l.getBetaMoves()}\n")
        else:
            if not self.l.native.hasNativeStructure(stats.Ntot):
                print("ERROR energy below native energy")
                print("Mapped stats:")
                stats.printCout()
                print("Lattice stats:")
                self.l.stats.printCout()
                
                if self.l.stats.Etot == stats.Etot:
                    self.l.writePDBMultiChain("belowNativeEnergy.pdb")
                    sys.exit(1)
    
    def printGnuPlotDataWR(self, filename: str, num_procs: int, my_rank: int):
        with open(filename, 'w') as f:
            print(f"printing to {filename}")
            for beta_id in range(num_procs):
                f.write(f"BETA\t{100.0 * self.l.betas[beta_id]}\n")
                for q in range(QMAX):
                    if self.energy_map[beta_id][q] > 0:
                        f.write(f"{q}\t{self.energy_map[beta_id][q]}\n")
            print(f"finished printing to {filename}")
    
    def printEMap(self, filename: str, num_procs: int, my_rank: int):
        with open(filename, 'w') as f:
            print(f"printing to {filename}")
            for beta_id in range(num_procs):
                f.write(f"BETA\t{100.0 * self.l.betas[beta_id]}\n")
                for energy, count in self.e_map[beta_id].items():
                    f.write(f"{energy / 100.0}\t{count}\n")
    
    def printHBondMap(self, filename: str, num_procs: int, my_rank: int):
        with open(filename, 'w') as f:
            print(f"printing to {filename}")
            for beta_id in range(num_procs):
                f.write(f"BETA\t{100.0 * self.l.betas[beta_id]}\n")
                for q in range(QMAX):
                    if self.hbond_map[beta_id][q] > 0:
                        f.write(f"{q}\t{self.hbond_map[beta_id][q]}\n")
            print(f"finished printing to {filename}")
